### PR TITLE
Update Replica count too while ensuring deployment

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
SSIA

### CODE REVIEW

File: lib/resourcemerge/apps.go [hunk 1]
```diff
@@ -10,6 +10,11 @@ import (
 func EnsureDeployment(modified *bool, existing *appsv1.Deployment, required appsv1.Deployment) {
 	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
 
+	if required.Spec.Replicas != nil && required.Spec.Replicas != existing.Spec.Replicas {
```
Comment (abhinavdahiya) [inline lib/resourcemerge/apps.go]: Why do you think this is important?

Comment (vikaschoudhary16) [reply lib/resourcemerge/apps.go]: We are using this lib in machine-api-operator. When MAO reconciles its owned deployments, it is desired that replica count should also be reconciled. https://jira.coreos.com/browse/CLOUD-484

File: lib/resourcemerge/apps_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	expectedReplicas := int32(2)
+	labelSelector := metav1.LabelSelector{}
+	tests := []struct {
+		name     string
+		existing appsv1.Deployment
+		required appsv1.Deployment
+
+		expectedModified bool
+		expected         appsv1.Deployment
+	}{{
```
Comment (abhinavdahiya) [inline lib/resourcemerge/apps_test.go]: this is different in format.. than others

**Comment by wking** [comment]
`replicas` seems like something that we'll want to be able to manage via release images.  In fact, the fewer knobs we leave *unmanaged*, the happier I'll be, so I think the burden of proof should be on "why *shouldn't* we manage this?".

/approve

I'll wait for @abhinavdahiya to weigh in on whether that makes sense to him and whether he wants [a different format for the unit tests][1].

[1]: https://github.com/openshift/cluster-version-operator/pull/198#discussion_r292244502

File: lib/resourcemerge/apps_test.go (hunk 2 above)
Comment (vikaschoudhary16) [reply lib/resourcemerge/apps_test.go]: Can you please explain a bit what you mean by "other" format? I tried to follow the format of the tests recently got added in[ #99](https://github.com/openshift/cluster-version-operator/pull/199/files#diff-415c13f11ffc32696c5d69b900b3fe58R11). Also i checked couple more existing unit test files like status_test.go and sync_test.go, those too seem to have different formats.
I think i did not understand correctly what you refering to as "others".

**Comment by abhinavdahiya** [comment]
> `replicas` seems like something that we'll want to be able to manage via release images. In fact, the fewer knobs we leave _unmanaged_, the happier I'll be, so I think the burden of proof should be on "why _shouldn't_ we manage this?".
> 
> /approve
> 
> I'll wait for @abhinavdahiya to weigh in on whether that makes sense to him and whether he wants [a different format for the unit tests](https://github.com/openshift/cluster-version-operator/pull/198#discussion_r292244502).

For second level operators that manage specific operands, it is very justifiable to say i own my operand and no changes are allowed, because the scope is very well defined..

For CVO because we manage the operators, my goal was to make sure we have a way to define _I care about this and I will enforce this, but you can maybe add new things if we feel strongly about it_ And that's why "manage all and prove why i it is un managed" is not a great basis for CVO...

As for replicas, most operators have no replicas defined, ie they get 1 replica using the server-side defaults. But i might be useful for user envs to make more than one operator replica running for more resiliency... So user might scale the operator deployment without requiring them to go to the unsupported bucket of Overrides.

But, an option is that if the operator team decides to put replicas 1 in the manifest, we enforce it. Initially I didn't go this route too, as most operators might define replicas _not really_ wanting that behavior.. And there was no need for CVO to enforce replicas as ClusterOperator check is the backstop for the business logic.

**Comment by wking** [comment]
> Initially I didn't go this route too, as most operators might define replicas *not really* wanting that behavior.

I'm fine making those operator bugs, since "enforce `replicas` when the release-manifest sets it" allows the manifest owner to decide if they care about `replicas` or not.

**Comment by vikaschoudhary16** [comment]
@wking @abhinavdahiya Ping!!

**Pull request merged**